pkg/types/types.go

```diff
@@ -197,29 +197,59 @@ type Metric struct {
 }
 
 // MergeMetrics merges metrics by name.
-func MergeMetrics(metrics [][]Metric) []Metric {
+// It returns merged metrics, number of rendered data points for the returned metrics,
+// and number of inconsistent data points seen (if consistencyCheck is true).
+func MergeMetrics(metrics [][]Metric, consistencyCheck bool) ([]Metric, int, int) {
 	if len(metrics) == 0 {
-		return nil
+		return nil, 0, 0
 	}
 
 	if len(metrics) == 1 {
-		return metrics[0]
+		pointCount := 0
+		ms := metrics[0]
+		for _, m := range ms {
+			pointCount += len(m.Values)
+		}
+		return ms, pointCount, 0
 	}
 
-	names := make(map[string][]Metric)
+	metricByNames := make(map[string][]Metric)
 
 	for _, ms := range metrics {
 		for _, m := range ms {
-			names[m.Name] = append(names[m.Name], m)
+			metricByNames[m.Name] = append(metricByNames[m.Name], m)
 		}
 	}
 
 	merged := make([]Metric, 0)
-	for _, ms := range names {
-		merged = append(merged, mergeMetrics(ms))
+	pointCount := 0
+	inconsistencyCount := 0
+	inconsMetricReportLimit := 10
+	var inconsMetricReport []map[string]interface{}
+	for _, ms := range metricByNames {
+		m, c := mergeMetrics(ms, consistencyCheck)
+		if c > 0 && len(inconsMetricReport) < inconsMetricReportLimit {
+			inconsMetricReport = append(inconsMetricReport, map[string]interface{}{
+				"metricName":         m.Name,
+				"start":              m.StartTime,
+				"stop":               m.StopTime,
+				"step":               m.StepTime,
+				"inconsistentPoints": c,
+			})
+		}
+		merged = append(merged, m)
+		inconsistencyCount += c
+		pointCount += len(m.Values)
 	}
 
-	return merged
+	if consistencyCheck && inconsistencyCount > 0 {
+		corruptionLogger.Warn("metric consistency issue",
+			zap.Any("inconsistentMetric", inconsMetricReport),
+			zap.Int("totalInconsistencies", inconsistencyCount),
+		)
+	}
+
+	return merged, pointCount, inconsistencyCount
```

Comment:
I wonder if `pointCount` and `inconsistencyCount` should be calculated in this func and passed back to cmd/app.

An alternative approach would be to add inconsistency/replicaMismatch bool into struct Metric, assign it inside MergeMetrics, and have cmd/app calculate pointCount and inconsistencyCount by iterating over Metric structs.

@grzkv @bom-d-van wdyt?

Reply:
moving them into `struct Metric` does make the function interface cleaner. but I guess the aggregation and logging logics still need to live somewhere.

maybe in the future we can return the summary result as a struct if the interface continue to grow.

Reply:
So which approach do you think would be preferable? The current one or the one I described above?

Reply:
I think the `struct Metric` way would be nicer. Then we can have a separate function for logging and summary and requires no interface changes for the MergeMetrics function. But I will leave it for Emad to make the call.

Reply:
I totally agree that moving the mismatch counts to metric is cleaner. However, I kinda have the feeling that having `mismatch` on `metric` can be conceptually irrational because a single metric data points cannot have mismatch (it cannot be defined for a single metric data point) but data points of a group of metrics merged into one can. So I think maybe encapsulating Metric in a MergedMetric or something called like this would be suitable for what we are discussing here. Any thoughts?
But as we talked with @grzkv before, I think keeping it as simple as possible is benefitial.